File: src/algorithms/hash-maps/counting/ransom-note/sources/ransom-note.ts
// Ransom Note — check if a ransom note can be constructed from magazine characters
function ransomNote(ransomNoteText: string, magazine: string): boolean {
  const charCounts = new Map<string, number>(); // @step:initialize
  for (const currentChar of magazine) {
    charCounts.set(currentChar, (charCounts.get(currentChar) ?? 0) + 1); // @step:increment-count
  }
  for (const currentChar of ransomNoteText) {
    const updatedCount = (charCounts.get(currentChar) ?? 0) - 1; // @step:decrement-count
    if (updatedCount < 0) return false; // @step:complete
    charCounts.set(currentChar, updatedCount); // @step:decrement-count
  }
  return true; // @step:complete
}

File: src/algorithms/hash-maps/counting/ransom-note/sources/ransom-note.rs
// Ransom Note — check if a ransom note can be constructed from magazine characters
use std::collections::HashMap;

fn ransom_note(ransom_note_text: &str, magazine: &str) -> bool {
    let mut char_counts: HashMap<char, i32> = HashMap::new(); // @step:initialize
    for current_char in magazine.chars() {
        *char_counts.entry(current_char).or_insert(0) += 1; // @step:increment-count
    }
    for current_char in ransom_note_text.chars() {
        let updated_count = char_counts.entry(current_char).or_insert(0);
        *updated_count -= 1; // @step:decrement-count
        if *updated_count < 0 {
            return false; // @step:complete
        }
        // @step:decrement-count
    }
    true // @step:complete
}
